File: hydroc/mpi/solver.c
#include <math.h>
#include <mpi.h>
#include <stdlib.h>
#include <stdio.h>
#include "solver.h"
#include "allocmultid.h"

#define MAX(x,y) ((y) > (x) ? (y) : (x))

const float eosgamma=5./3.;

void bufferGuardcells(float ***u, const int nx, const int ny, 
                      const char xydim, 
                      MPI_Comm cartcomm, int left, int right) {
    int datasize;
    float *sendbuffer, *recvbuffer;
    int i,j,v,count;
    MPI_Status rstatus;
    const int lefttag = 1;
    const int righttag = 2;

    datasize = (ny-2*NGUARD)*NGUARD*NVARS;
    sendbuffer = (float *)malloc(datasize*sizeof(float));
    recvbuffer = (float *)malloc(datasize*sizeof(float));
    if (!sendbuffer || !recvbuffer) {
        fprintf(stderr,"Error mallocing send/recv buffer(!?)\n");
        return;
    }

    /* rightward */
    /* copy data into a buffer */

    /* MPI_Sendrecv(sendbuffer, datasize,... ?
                    recvbuffer, datasize,... ? */
    /* copy data out of a buffer */

    /* leftward */
    /* copy data into a buffer */

    /* MPI_Sendrecv(sendbuffer, datasize,... ?
                    recvbuffer, datasize,... ? */
    /* copy data out of a buffer */


    free(sendbuffer);
    free(recvbuffer);
}

void vectorGuardcells(float ***u, const int nx, const int ny, const char xydim,
                      MPI_Comm cartcomm, const int left, const int right) {
    const int lefttag=1, righttag=2;
    int ierr;
    MPI_Status rstatus;
    MPI_Datatype xbctype;

    /* ierr = MPI_Type_vector(...? */
    MPI_Type_commit(&xbctype);

    /* ierr = MPI_Sendrecv(...?  cartcomm, &rstatus); */

    /* ierr = MPI_Sendrecv(...?  cartcomm, &rstatus); */

    MPI_Type_free(&xbctype);
}

void periodicBCs(float ***u, const int nx, const int ny, const char xydim) {

    if (xydim == 'y') {
        int n = ny;
        for (int j=0; j<NGUARD; j++) {
            for (int i=0; i<nx; i++) { 
                for (int var=0 ; var<NVARS; var++) {
                    u[j][i][var] = u[n-2*NGUARD+j][i][var];
                    u[n-NGUARD+j][i][var] = u[j+NGUARD][i][var];
                }
            } 
        }
    } else if (xydim == 'x') {
        int n = nx;
        for (int j=0; j<ny; j++) { 
            for (int i=0; i<NGUARD; i++) { 
                for (int var=0 ; var<NVARS; var++) {
                    u[j][i][var] = u[j][n-2*NGUARD+i][var];
                    u[j][n-NGUARD+i][var] = u[j][i+NGUARD][var];
                }
            }
        } 
    }
}


float cfl(float ***u, const int nx, const int ny) {
   float c;
   float globalc;

   c = 0.;
   
   for (int j=0; j<ny; j++) {
       for (int i=0; i<nx; i++)  {
            float vx = fabsf(u[j][i][IMOMX]/u[j][i][IDENS]);
            float vy = fabsf(u[j][i][IMOMY]/u[j][i][IDENS]);
            float v = vx;
            if (vy > vx) v=vy;
            float p = (u[j][i][IENER] - 0.5*(vx*vx+vy*vy)*u[j][i][IDENS])*(eosgamma-1.);
            float cloc = v + sqrtf(fabsf(p*eosgamma/u[j][i][IDENS]));
            c = MAX(c,cloc);
        }
   }

   /* find global max of all cs */
   MPI_Allreduce(&c, &globalc, 1, MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
   return (1./globalc);
}

void initialconditions(float ***u,const int nx, const int ny, const int n, const int start) {
  int i, j;
  const float backgrounddens=1., projdens=50., projvel=4., p0=1.;
  float r;
  float x,y;
  
 
  for (j=0;j<ny;j++) {
      y = ((j-NGUARD)*1.-n/2.);
      for (i=0;i<nx;i++) {
          x = ((i-NGUARD+start)*1.-n/2.);
          r =sqrtf(x*x+y*y);

          if (r < 0.1*sqrtf(n*n*1.+n*n*1.)) {
            u[j][i][IDENS] = projdens;
            u[j][i][IMOMX] = projvel*projdens;
            u[j][i][IMOMY] = 0.;
            u[j][i][IENER] = 0.5*(projdens*projvel*projvel)+p0/(eosgamma-1.);
          } else {
            u[j][i][IDENS] = backgrounddens;
            u[j][i][IMOMX] = 0.;
            u[j][i][IMOMY] = 0.;
            u[j][i][IENER] = p0/(eosgamma-1.);
          }
      }
  }
}

void hydroflux(float **v, float *rc, float **u, const int n) {
  float c = 0.;
  
  for (int i=0; i<n; i++) {
    float vx = u[i][IMOMX]/u[i][IDENS];
    float p  = u[i][IENER] - 0.5*(u[i][IMOMX]*u[i][IMOMX]+u[i][IMOMY]*u[i][IMOMY])
                                   /u[i][IDENS];
    p *= (eosgamma-1.);
  
    v[i][IDENS] = u[i][IMOMX];
    v[i][IMOMX] = u[i][IMOMX]*vx + p;
    v[i][IMOMY] = u[i][IMOMY]*vx;
    v[i][IENER] = (u[i][IENER]+p)*vx;

    float newc = (fabs(vx)+sqrt(fabs(eosgamma*p/u[i][IDENS])));
    c = MAX((c),newc);
  }

  if (c > 0.) {
      for (int i=0; i<n; i++) {
          for (int var=0; var<NVARS; var++) {
            v[i][var] /= c;
          }
      }
  }

  *rc = c;
}


void tvd1d(float **u, const int n, const float dt) {
    float **v, **u1, **wr, **wl;
    float **dfl, **dfr, **flux;
    float c;
    const float smallp=1.e-3;

    v  = alloc2d_float(n, NVARS);
    
    hydroflux(v, &c, u, n);

    wl = alloc2d_float(n, NVARS);
    wr = alloc2d_float(n, NVARS);
    for (int i=0; i<n; i++) {
        for (int var=0; var<NVARS; var++) {
            wr[i][var] = u[i][var] + v[i][var];
            wl[i][var] = u[i][var] - v[i][var];
        }
    }

    flux = alloc2d_float(n, NVARS);
    for (int i=0; i<n; i++) {
        int irshift = i+1;
        if (irshift >= n) irshift -= n;
        for (int var=0; var<NVARS; var++) {
            flux[i][var] = c*(wr[i][var] - wl[irshift][var])/2.;
        }
    }
    
    u1 = alloc2d_float(n, NVARS);
    for (int i=0; i<n; i++) {
        int ilshift = i-1;
        if (ilshift < 0) ilshift += n;
        for (int var=0; var<NVARS; var++) {
            u1[i][var] = u[i][var] - (flux[i][var] - flux[ilshift][var])*dt/2.;
        }
    }
    
    hydroflux(v, &c, u1, n);
    for (int i=0; i<n; i++) {
        for (int var=0; var<NVARS; var++) {
            wr[i][var] = u1[i][var] + v[i][var];
            wl[i][var] = u1[i][var] - v[i][var];
        }
    }

    dfr = alloc2d_float(n, NVARS);
    for (int i=0; i<n; i++) {
        int irshift = i+1;
        if (irshift >= n) irshift -= n;
        int ilshift = i-1;
        if (ilshift < 0) ilshift += n;
        for (int var=0; var<NVARS; var++) {
            dfr[i][var] = 0.;
            float dfrp = c*(wr[irshift][var] - wr[i][var])/2.; 
            float dfrm = c*(wr[i][var] - wr[ilshift][var])/2.; 
            if (dfrp*dfrm > 0.) 
                dfr[i][var] = (2.*(dfrp*dfrm)/(dfrp+dfrm));
        }
    }
    
    dfl = alloc2d_float(n, NVARS);
    for (int i=0; i<n; i++) {
        int irshift = i+1;
        if (irshift >= n) irshift -= n;
        int irrshift = i+2;
        if (irrshift >= n) irrshift -= n;
        for (int var=0; var<NVARS; var++) {
            dfl[i][var] = 0.;
            float dflp = c*(wl[irshift][var] - wl[irrshift][var])/2.; 
            float dflm = c*(wl[i][var] - wl[irshift][var])/2.; 
            if (dflp*dflm > 0.) 
                dfl[i][var] = (2.*(dflp*dflm)/(dflp+dflm));
        }
    }
    
    for (int i=0; i<n; i++) {
        int irshift = i+1;
        if (irshift >= n) irshift -= n;
        for (int var=0; var<NVARS; var++) {
            flux[i][var] = (c*(wr[i][var] - wl[irshift][var]) 
                             + (dfr[i][var]-dfl[i][var]))/2.;
        }
    }
    

    for (int i=NGUARD; i<n-NGUARD; i++) {
        int ilshift = i-1;
        if (ilshift < 0) ilshift += n;
        for (int var=0; var<NVARS; var++) {
            u[i][var] -= (flux[i][var] - flux[ilshift][var])*dt;
        }
        float rho=u[i][IDENS];
        float vx=u[i][IMOMX]/rho;
        float vy=u[i][IMOMY]/rho;
        if ( (u[i][IENER] - 0.5*(vx*vx+vy*vy)*rho)*(eosgamma-1.) < smallp ) 
            u[i][IENER] = smallp/(eosgamma-1.) +  0.5*(vx*vx+vy*vy)*rho ;
    }
    
    free2d_float(v);
    free2d_float(u1);
    free2d_float(wr);
    free2d_float(wl);
    free2d_float(dfl);
    free2d_float(dfr);
    free2d_float(flux);
    return;
}

void xytranspose(float ***ut, float ***u, const int nx, const int ny) {
    int i,j;
    for (i=0; i<nx; i++) {
        for (j=0; j<ny; j++) {
            ut[i][j][IDENS] = u[j][i][IDENS];
            ut[i][j][IENER] = u[j][i][IENER];
            ut[i][j][IMOMX] = u[j][i][IMOMY];
            ut[i][j][IMOMY] = u[j][i][IMOMX];
        }
    }
}


void xsweep(float ***u, const int nx, const int ny, const float dt){
  for (int j=0; j<ny; j++) {
     tvd1d(u[j],nx,dt);
  }
}

void timestep(float ***u, const int nx, const int ny, float *dt, 
              MPI_Comm cartcomm, int left, int right) {
    float ***ut;

    ut = alloc3d_float(nx, ny, NVARS);
    *dt=0.25*cfl(u,nx,ny);

    /* the x sweep */
    periodicBCs(u,nx,ny,'x');
    //should be bufferGuardcells(u,nx,ny,'x',cartcomm,left,right);
    xsweep(u,nx,ny,*dt);

    /* the y sweeps */
    xytranspose(ut,u,nx,ny);
    periodicBCs(ut,ny,nx,'x');
    xsweep(ut,ny,nx,*dt);
    periodicBCs(ut,ny,nx,'x');
    xsweep(ut,ny,nx,*dt);

    /* 2nd x sweep */
    xytranspose(u,ut,ny,nx);
    periodicBCs(u,nx,ny,'x');
    //should be bufferGuardcells(u,nx,ny,'x',cartcomm,left,right);
    xsweep(u,nx,ny,*dt);

    free3d_float(ut,nx);
}

File: hydrof/mpi/solver.f90
module solver
  use mpi
  implicit none

  integer, parameter :: idens=1, iener=4, imomx=2, imomy=3, nvars=4
  integer, parameter :: nguard = 2
  real, parameter :: gamma=5./3.

  public :: initialconditions, timestep, nguard, gamma
  private:: cfl, xytranspose, xsweep, tvd1d, hydroflux

contains

subroutine periodicBCs(u,xydim)
    implicit none
    real, dimension(:,:,:), intent(INOUT) :: u
    character, intent(IN) :: xydim
    integer :: n

    if (xydim == 'y') then
        n = size(u,3)
        u(:,:,1:2)   = u(:,:,n-3:n-2)
        u(:,:,n-1:n) = u(:,:,3:4)
    else if (xydim == 'x') then
        n = size(u,2)
        u(:,1:2,:)   = u(:,n-3:n-2,:)
        u(:,n-1:n,:) = u(:,3:4,:)
    endif
end subroutine periodicBCs

subroutine bufferGuardcells(u,xydim,commcart,left,right)
    real, dimension(:,:,:), intent(INOUT) :: u
    character, intent(IN) :: xydim
    integer, intent(IN) :: commcart,left,right
    integer :: nx, ny
    integer :: datasize
    real, dimension(:), allocatable:: sendbuffer, rcvbuffer
    integer :: ierr
    integer :: righttag, lefttag
    integer, dimension(MPI_STATUS_SIZE) :: rstatus

    nx = size(u,2) 
    ny = size(u,3) 

    lefttag = 1
    righttag = 2

    datasize = (ny-2*nguard)*nguard*nvars
    allocate(sendbuffer(datasize),rcvbuffer(datasize))

    ! rightward
    ! Copy data into buffer
    !
    ! call MPI_Sendrecv(sendbuffer, datasize,  ..?
    !                   rcvbuffer,  datasize,  ..?
    ! copy data out of buffer.

    ! leftward
    ! Copy data into buffer
    !
    ! call MPI_Sendrecv(sendbuffer, datasize,  ..?
    !                   rcvbuffer,  datasize,  ..?
    ! copy data out of buffer.


    deallocate(sendbuffer,rcvbuffer)
end subroutine bufferGuardcells

subroutine vectorGuardcells(u,xydim,commcart,left,right)
    real, dimension(:,:,:), intent(INOUT) :: u
    character, intent(IN) :: xydim
    integer, intent(IN) :: commcart, left, right
    integer :: nx, ny
    integer :: xbctype
    integer :: lefttag, righttag
    integer :: ierr
    integer, dimension(MPI_STATUS_SIZE) :: rstatus

    nx = size(u,2) 
    ny = size(u,3) 

    lefttag = 1
    righttag = 2

    ! call MPI_Type_vector(...?)
    call MPI_Type_commit(xbctype, ierr)

    ! call MPI_Sendrecv(...?
    ! call MPI_Sendrecv(...?

    call MPI_Type_free(xbctype, ierr)
end subroutine vectorGuardcells

subroutine timestep(u,dt,commcart,left,right)
    implicit none
    real, dimension(:,:,:), intent(INOUT) :: u
    real, intent(OUT) :: dt
    integer, intent(in) :: commcart,left,right

    real, dimension(nvars,size(u,3),size(u,2)) :: ut

    dt=0.25*cfl(u)
! the x sweep
    call periodicBCs(u,'x')
    !call vectorGuardcells(u,'x',commcart,left,right)
    call xsweep(u,dt)
! the y sweeps
    call xytranspose(ut,u)
    call periodicBCs(ut,'x')
    call xsweep(ut,dt)
    call periodicBCs(ut,'x')
    call xsweep(ut,dt)
! 2nd x sweep
    call xytranspose(u,ut)
    call periodicBCs(u,'x')
    !call vectorGuardcells(u,'x',commcart,left,right)
    call xsweep(u,dt)
end subroutine timestep

real function cfl(u)
   implicit none
   real, intent(IN) :: u(:,:,:)
   real, dimension(size(u,2),size(u,3)) :: v, p
   real :: c
   real :: globc
   integer :: ierr

   v = abs(u(imomx,:,:)/u(idens,:,:))
   where (v < abs(u(imomy,:,:))/u(idens,:,:))
      v = abs(u(imomy,:,:)/u(idens,:,:))
   endwhere

   p = (u(iener,:,:)-0.5*(u(imomx,:,:)**2+u(imomy,:,:)**2)/u(idens,:,:))*(gamma-1.)
   c = maxval(v+sqrt(abs((gamma*p))/u(idens,:,:))) 
   
   ! find global max of all c's
   call MPI_Allreduce(c, globc, 1, MPI_REAL, MPI_MAX, MPI_COMM_WORLD, ierr)
   cfl=1./globc
end function cfl

subroutine initialconditions(u,n,startn)
  implicit none
  real, intent(OUT)   :: u(:,:,:)
  integer, intent(IN) :: n, startn
  integer :: nx, ny
   
  integer :: i,j
  real, parameter :: backgrounddens=1., projdens=50., projvel=4., p0=1.
  real :: r(size(u,2), size(u,3))

  nx = size(u,2)
  ny = size(u,3)
 
  forall (i=1:nx,j=1:ny)
      r(i,j) =sqrt((j*1-n/2.)**2. + ((i+startn-1)*1.-n/2.)**2)
  endforall

  where (r < 0.1*sqrt(n*n+n*n*1.))
    u(idens,:,:) =projdens
    u(imomx,:,:) =projdens*projvel
    u(imomy,:,:) =0
    u(iener,:,:) =0.5*(projdens*projvel*projvel)+p0/(gamma-1.)
  elsewhere
    u(idens,:,:) =backgrounddens
    u(imomx,:,:) =0.
    u(imomy,:,:) =0.
    u(iener,:,:) =p0/(gamma-1.)
  endwhere

end subroutine initialconditions

subroutine xsweep(u,dt)
  implicit none
  real, intent(INOUT), dimension(:,:,:) :: u
  real, intent(IN) :: dt
  integer :: j

  do j=1,size(u,3)
     call tvd1d(u(:,:,j),dt)
  enddo
end subroutine xsweep

subroutine tvd1d(u,dt)
  implicit none
  real, intent(IN) :: dt
  real, dimension(:,:), intent(INOUT) :: u

  real, dimension(size(u,1),size(u,2)) :: v, u1, wr, wl
  real, dimension(size(u,1),size(u,2)) :: dfrp, dfrm, dflm, dflp
  real, dimension(size(u,1),size(u,2)) :: dfl, dfr
  real, dimension(size(u,1),size(u,2)) :: flux
  real, dimension(size(u,1),size(u,2)) :: du
  real :: c
  real, parameter :: smallp = 1.e-3
  integer :: i, n

  n = size(u,2)

  call hydroflux(v,c,u)
  wr = u+v
  wl = u-v
  flux=c*(wr - eoshift(wl,1,dim=2))/2.
  u1 = u - (flux - eoshift(flux,-1,dim=2))*dt/2.

  call hydroflux(v,c,u1)
  wr = u1+v
  wl = u1-v

  dfrp = c*(eoshift(wr,1,dim=2)-wr)/2.
  dfrm = c*(wr-eoshift(wr,-1,dim=2))/2.
  where (dfrp*dfrm > 0) 
     dfr = 2*dfrp*dfrm/(dfrp+dfrm)
  elsewhere
     dfr = 0
  endwhere
  
  dflp = c*(eoshift(wl,1,dim=2)-eoshift(wl,2,dim=2))/2.
  dflm = c*(wl-eoshift(wl,1,dim=2))/2.
  where (dflp*dflm > 0) 
     dfl = 2*dflp*dflm/(dflp+dflm)
  elsewhere
     dfl = 0
  endwhere
  
  flux = (c*(wr-eoshift(wl,1,dim=2)) + (dfr-dfl))/2.
  du = (flux-eoshift(flux,-1,dim=2))*dt
  u(:,3:n-2) = u(:,3:n-2) - du(:,3:n-2)

  do i=1,n
    if ( ((u(iener,i)-0.5*(u(imomx,i)**2+u(imomy,i)**2)/u(idens,i))*(gamma-1.)) < smallp ) &
              u(iener,i) = 0.5*(u(imomx,i)**2+u(imomy,i)**2)/u(idens,i) + smallp/(gamma-1.)
  enddo
end subroutine tvd1d

subroutine hydroflux(v,c,u)
  implicit none
  real, intent(OUT)    :: c
  real, dimension(:,:), intent(IN) :: u
  real, dimension(:,:), intent(OUT) :: v
  real, dimension(size(u,2)) :: vx, p

  c  = 0.
  vx = u(imomx,:)/u(idens,:)
  p  = (u(iener,:)-(u(imomx,:)**2+u(imomy,:)**2)/u(idens,:)/2)*(gamma-1)

  v(idens,:)=u(imomx,:)
  v(imomx,:)=u(imomx,:)*vx + p
  v(imomy,:)=u(imomy,:)*vx
  v(iener,:)=(u(iener,:)+p)*vx

  c=maxval(abs(vx)+sqrt(abs(gamma*p/u(idens,:))))
  if (c > 0) v = v/c
end subroutine hydroflux

subroutine xytranspose(ut,u)
  implicit none
  real, dimension(:,:,:), intent(IN) :: u
  real, dimension(:,:,:), intent(OUT) :: ut

  ut(idens,:,:)=transpose(u(idens,:,:))
  ut(imomx,:,:)=transpose(u(imomy,:,:))
  ut(imomy,:,:)=transpose(u(imomx,:,:))
  ut(iener,:,:)=transpose(u(iener,:,:))
end subroutine xytranspose

end module solver
